tree: handle statement fingerprints as HintInjectionDonors

michae2 opened the pull request:
Make changes to HintInjectionDonor in order to support using statement fingerprints as the donor. (See individual commits for details.)

Informs: #153633

Release note: None

michae2 committed:
tree: also visit Statement nodes in ExtendedVisitor

Similar to 300e81bf0e170c0f2408df19e7e784c4911ccc63, we also need to
visit Statement nodes during hint injection, so add them to
ExtendedVisitor. This commmit adds VisitStatementPre and
VisitStatementPost to ExtendedVisitor and extendedSimpleVisitor.

It also converts debugVisitor from a Visitor to an ExtendedVisitor which
will provide more information. (ExprDebugString is currently only called
in one place, but I have some ideas for how to use these in the future.)

Informs: #153633

Release note: None

michae2 committed:
tree: handle statement fingerprints as HintInjectionDonors

The hint injection code added in #153689 could handle regular statements
as hint injection donors, but we also need to handle statement
*fingerprints* as hint injection donors. (In fact, we expect this will
be the common case.)

Statement fingerprints have several differences from regular SQL
statements:

- Constants and placeholders are replaced with _, which parse as
  `UnresolvedName("_")`.
- Tuples, Arrays, and ValuesClauses could be collapsed to `(_)` or
  `(_, __more__)` depending on how many items they contain.
- ValuesClauses could also look like `((1, 2, __more__), (__more__))`

We need to be able to match a donor statement fingerprint against a
regular target statement during hint injection. To do so, this commit
makes several changes:

1. A "collapsed" stack of booleans is maintained while walking the ASTs
   to track whether the current Tuple, Array, or ValuesClause in the
   donor has been collapsed into `__more__`. When this is true, we skip
   over the rest of the nodes until the end of the current Tuple, Array,
   or ValuesClause.
2. We now skip ParenExpr in the AST walks, since they're sometimes added
   to statement fingerprints.
3. `_` and `__more__` now act as wildcards during the target statement
   AST walk, matching any node. (And then after `__more__` we mark the
   current list as collapsed.)
4. JoinType of INNER (e.g. the INNER in INNER LOOKUP JOIN) needed
   special handling, because it's required when a join hint is added,
   but it is optional without a hint. To make validation work, we now
   elide INNER when FmtHideHints is set. But we also carry over an
   optional INNER from donor statement to target statement even when
   there is no accompanying join hint, which seems harmless but makes
   TestRandomInjectHints possible. The validation step should prevent
   accidental injection of a LEFT or other outer join type that wasn't
   present in the target statement.

Finally, testing has been updated to use statement fingerprints as the
hint injection donors.

Informs: #153633

Release note: None

michae2 committed:
tree: change hint injection validation to take FmtFlags arg

Instead of passing around settings everywhere, let the outer code look
up the value of `sql.stats.statement_fingerprint.format_mask` and
instead pass FmtFlags into the hint injection code. This will simplify
the next PR.

Informs: #153633

Release note: None

michae2 committed:
tree: also remove existing hints from target during hint injection

During hint injection, also remove existing hints from the target
statement if they are not present in the donor statement. This makes it
easier to understand the rewrite performed by hint injection, as well as
makes it possible for hint injection to remove hints.

Informs: #153633

Release note: None

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

> @mgartner reviewed 1 of 1 files at r1, 3 of 3 files at r2, 2 of 2 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-:-OdyD0eB7AX8PmovTplp:bcmw7ss)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`-- commits` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-Ody7k-HD2FfRuMxFQQj:-Ody7k-HD2FfRuMxFQQk:b-g0znnz):*
Should these match _any_ node or just nodes representing constant values?
___
*[`pkg/sql/sem/tree/inject_hints_test.go` line 213 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-OdyA37vD3ERJyS3iCEN:-OdyA37vD3ERJyS3iCEO:b-k5sc8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdafeecdc25fbd42891961b37a884c9d4b3b58f5/pkg/sql/sem/tree/inject_hints_test.go#L213)):*
> ```Go
> 			originalSQL:   "SELECT * FROM a WHERE b > 1 AND b < 10 ORDER BY c LIMIT 2",
> 			donorSQL:      "SELECT * FROM a@a_b_idx WHERE (b > _) AND (b < _) ORDER BY c LIMIT _",
> 			expectedSQL:   "SELECT * FROM a@a_b_idx WHERE (b > 1) AND (b < 10) ORDER BY c LIMIT 2",
> ```

nit: might be nice to have a test case with nested parens like `((b > 1) AND (b < 10))`
___
*[`pkg/sql/sem/tree/select.go` line 503 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-OdyCXqZ6Ql9vNHqN-58:-OdyCXqZ6Ql9vNHqN-59:b-4fmilh) ([raw file](https://github.com/cockroachdb/cockroach/blob/532c55bd92ca7b2125ff95bd0b4ee8be5b69a602/pkg/sql/sem/tree/select.go#L503)):*
> ```Go
> 		slices.Equal(ih.ZigzagIndexIDs, other.ZigzagIndexIDs) &&
> 		((ih.FamilyID == nil && other.FamilyID == nil) ||
> 			(ih.FamilyID != nil && other.FamilyID != nil && *ih.FamilyID == *other.FamilyID))
> ```

nit: if you separate the comparable fields of the struct into a sub-struct, you can simplify this a bit:

```diff
diff --git a/pkg/sql/sem/tree/select.go b/pkg/sql/sem/tree/select.go
index 89dd30214f9..1f81f57ba7e 100644
--- a/pkg/sql/sem/tree/select.go
+++ b/pkg/sql/sem/tree/select.go
@@ -16,6 +16,7 @@ package tree
 
 import (
 	"fmt"
+	"slices"
 
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
@@ -322,6 +323,16 @@ type FamilyID = catid.FamilyID
 //
 // It is used optionally after a table name in SELECT statements.
 type IndexFlags struct {
+	comparableIndexFlags
+	ZigzagIndexes  []UnrestrictedName
+	ZigzagIndexIDs []IndexID
+
+	// Restrict select to the specified column family.
+	// Used by changefeed.
+	FamilyID *FamilyID
+}
+
+type comparableIndexFlags struct {
 	Index   UnrestrictedName
 	IndexID IndexID
 	// Direction of the scan, if provided. Can only be set if
@@ -350,13 +361,15 @@ type IndexFlags struct {
 	// ZigzagIndexes means we saw TABLE@{FORCE_ZIGZAG=name}
 	// ZigzagIndexIDs means we saw TABLE@{FORCE_ZIGZAG=[ID]}
 	// The only allowable combinations are when a valid id and name are combined.
-	ForceZigzag    bool
-	ZigzagIndexes  []UnrestrictedName
-	ZigzagIndexIDs []IndexID
+	ForceZigzag bool
+}
 
-	// Restrict select to the specified column family.
-	// Used by changefeed.
-	FamilyID *FamilyID
+func (ih *IndexFlags) Equal(other *IndexFlags) bool {
+	return ih.comparableIndexFlags == other.comparableIndexFlags &&
+		slices.Equal(ih.ZigzagIndexes, other.ZigzagIndexes) &&
+		slices.Equal(ih.ZigzagIndexIDs, other.ZigzagIndexIDs) &&
+		((ih.FamilyID == nil && other.FamilyID == nil) ||
+			(ih.FamilyID != nil && other.FamilyID != nil && *ih.FamilyID == *other.FamilyID))
 }
 
 // ForceIndex returns true if a forced index was specified, either using a name
```


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-:-Oe2qQ3nDd6x5M5fuUmM:b-814g8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mgartner)

___
*[`-- commits` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-Ody7k-HD2FfRuMxFQQj:-Oe2qRCl6KiU2g4mRUtE:b-w80c1j):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Should these match _any_ node or just nodes representing constant values?
</blockquote></details>

You're right, they should only match nodes representing constant values / placeholders. (Or, I suppose, if someone used `_` or `__more__` as a column name, they should match that as well.)

But we do have another layer of match verification (`(*HintInjectionDonor).Validate` which compares SQL strings after formatting both donor and target as fingerprints) which should catch if a different node matches the _.

The checks during the `hintInjectionVisitor` AST walk are mostly sanity checks to make sure we're lining up the hint-donating and hint-receiving AST nodes correctly. And it simplifies `hintInjectionVisitor` quite a bit to allow these to match any node. For those reasons I'm inclined to leave it the way it is, if that sounds alright.
___
*[`pkg/sql/sem/tree/select.go` line 503 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-OdyCXqZ6Ql9vNHqN-58:-Oe2ui5Q0cdt6BhBBORJ:b-2gmwbx) ([raw file](https://github.com/cockroachdb/cockroach/blob/532c55bd92ca7b2125ff95bd0b4ee8be5b69a602/pkg/sql/sem/tree/select.go#L503)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: if you separate the comparable fields of the struct into a sub-struct, you can simplify this a bit:

```diff
diff --git a/pkg/sql/sem/tree/select.go b/pkg/sql/sem/tree/select.go
index 89dd30214f9..1f81f57ba7e 100644
--- a/pkg/sql/sem/tree/select.go
+++ b/pkg/sql/sem/tree/select.go
@@ -16,6 +16,7 @@ package tree
 
 import (
 	"fmt"
+	"slices"
 
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
@@ -322,6 +323,16 @@ type FamilyID = catid.FamilyID
 //
 // It is used optionally after a table name in SELECT statements.
 type IndexFlags struct {
+	comparableIndexFlags
+	ZigzagIndexes  []UnrestrictedName
+	ZigzagIndexIDs []IndexID
+
+	// Restrict select to the specified column family.
+	// Used by changefeed.
+	FamilyID *FamilyID
+}
+
+type comparableIndexFlags struct {
 	Index   UnrestrictedName
 	IndexID IndexID
 	// Direction of the scan, if provided. Can only be set if
@@ -350,13 +361,15 @@ type IndexFlags struct {
 	// ZigzagIndexes means we saw TABLE@{FORCE_ZIGZAG=name}
 	// ZigzagIndexIDs means we saw TABLE@{FORCE_ZIGZAG=[ID]}
 	// The only allowable combinations are when a valid id and name are combined.
-	ForceZigzag    bool
-	ZigzagIndexes  []UnrestrictedName
-	ZigzagIndexIDs []IndexID
+	ForceZigzag bool
+}
 
-	// Restrict select to the specified column family.
-	// Used by changefeed.
-	FamilyID *FamilyID
+func (ih *IndexFlags) Equal(other *IndexFlags) bool {
+	return ih.comparableIndexFlags == other.comparableIndexFlags &&
+		slices.Equal(ih.ZigzagIndexes, other.ZigzagIndexes) &&
+		slices.Equal(ih.ZigzagIndexIDs, other.ZigzagIndexIDs) &&
+		((ih.FamilyID == nil && other.FamilyID == nil) ||
+			(ih.FamilyID != nil && other.FamilyID != nil && *ih.FamilyID == *other.FamilyID))
 }
 
 // ForceIndex returns true if a forced index was specified, either using a name
```
</blockquote></details>

Ah nice, I was wondering about this.

I tried it out, but unfortunately it requires changing all uses of literal IndexFlags from:

```
tree.IndexFlags{Index: ...}
```

to:

```
tree.IndexFlags{ComparableIndexFlags: tree.ComparableIndexFlags{Index: ...}}
```

which seems like kind of an abstraction leak. So for now I've left it the way it was.
___
*[`pkg/sql/sem/tree/inject_hints_test.go` line 213 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-OdyA37vD3ERJyS3iCEN:-Oe2vKzR5Os0rUf7nSf0:b-48vilg) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdafeecdc25fbd42891961b37a884c9d4b3b58f5/pkg/sql/sem/tree/inject_hints_test.go#L213)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: might be nice to have a test case with nested parens like `((b > 1) AND (b < 10))`
</blockquote></details>

Good call, done.


<!-- Sent from Reviewable.io -->

michae2 commented:
Claude points out that the collapsed stack is not concurrency-safe, so InjectHints can't be called from multiple goroutines. I'm going to rework it slightly to be concurrency-safe.

michae2 committed:
tree: ensure that InjectHints is thread-safe

By re-using the same `collapsed` stack for both walks, InjectHints is
not safe to call from multiple goroutines. We need to be able to safely
call InjectHints from multiple sessions for the same donor.

To fix this, use separate `collapsed` stacks for both walks.

Informs: #153633

Release note: None

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @mgartner reviewed 1 of 3 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-:-OeHYhva8-6BRKoDinqO:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`-- commits` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-Ody7k-HD2FfRuMxFQQj:-OeHXvyi55mNFQvQkhyC:b-i5fhz6):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

You're right, they should only match nodes representing constant values / placeholders. (Or, I suppose, if someone used `_` or `__more__` as a column name, they should match that as well.)

But we do have another layer of match verification (`(*HintInjectionDonor).Validate` which compares SQL strings after formatting both donor and target as fingerprints) which should catch if a different node matches the _.

The checks during the `hintInjectionVisitor` AST walk are mostly sanity checks to make sure we're lining up the hint-donating and hint-receiving AST nodes correctly. And it simplifies `hintInjectionVisitor` quite a bit to allow these to match any node. For those reasons I'm inclined to leave it the way it is, if that sounds alright.
</blockquote></details>

I see. So if my donor statement is `SELECT _ FROM t@idx` then during the AST it would match a statement like `SELECT (SELECT s FROM s LIMIT 1) FROM t`, but would later fail before executing with the hints due to the difference in fingerprint formatting? Do you think formatting the new AST on every execution (is it every execution) adds a lot of overhead? Or are we already formatting all statements (if we are, then I am concerned about that...).
___
*[`pkg/sql/sem/tree/select.go` line 503 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-OdyCXqZ6Ql9vNHqN-58:-OeHX7iX0_Oq9RwACyMR:b-j656ur) ([raw file](https://github.com/cockroachdb/cockroach/blob/532c55bd92ca7b2125ff95bd0b4ee8be5b69a602/pkg/sql/sem/tree/select.go#L503)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Ah nice, I was wondering about this.

I tried it out, but unfortunately it requires changing all uses of literal IndexFlags from:

```
tree.IndexFlags{Index: ...}
```

to:

```
tree.IndexFlags{ComparableIndexFlags: tree.ComparableIndexFlags{Index: ...}}
```

which seems like kind of an abstraction leak. So for now I've left it the way it was.
</blockquote></details>

Oh darn, I didn't realize that. Thanks for trying!


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> This is great!

> @DrewKimball reviewed 1 of 1 files at r1, 3 of 3 files at r5, 3 of 3 files at r6, 2 of 2 files at r7, 3 of 3 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-:-OeHooOA5cmhS6bgBOWE:b48xzcn)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/sem/tree/inject_hints.go` line 42 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-OeHoYAA1T1RMQ03k0yb:-OeHoYAA1T1RMQ03k0yc:b-hb1ddu) ([raw file](https://github.com/cockroachdb/cockroach/blob/9578f8330dfe2f6fddba633486d3a39415e6aca1/pkg/sql/sem/tree/inject_hints.go#L42)):*
> ```Go
> 	hd := &HintInjectionDonor{
> 		ast:           ast,
> 		validationSQL: FormatStatementHideConstants(ast, fingerprintFlags, FmtHideHints),
> ```

Maybe we should do this in the hint cache and store the result once? A TODO is probably OK for now, though.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Thanks for the reviews!

bors r=mgartner,DrewKimball

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-:-OeJYjJm7v1P8lGjxgmx:br71zdf)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`-- commits` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-Ody7k-HD2FfRuMxFQQj:-OeIGv8zBKp2DHlZg9Bu:b-sv1trp):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I see. So if my donor statement is `SELECT _ FROM t@idx` then during the AST it would match a statement like `SELECT (SELECT s FROM s LIMIT 1) FROM t`, but would later fail before executing with the hints due to the difference in fingerprint formatting? Do you think formatting the new AST on every execution (is it every execution) adds a lot of overhead? Or are we already formatting all statements (if we are, then I am concerned about that...).
</blockquote></details>

That's right. Each query execution will first call `Validate` and then will call `InjectHints` if validation passes.

We are already formatting the AST at least twice when we prepare it, [here](https://github.com/cockroachdb/cockroach/blob/1b9970ee90ad3e405ce62860000f54d77d8a96cf/pkg/sql/statement.go#L81-L82). Validation will add a third formatting pass (and then another walk to rewrite the AST).

I was curious about the overhead, so I tried benchmarking this change calling FormatStatementHideConstants 100 times during prepare:

```
diff --git a/pkg/sql/statement.go b/pkg/sql/statement.go
index 92ea60fa9f6..fcc3969b2a5 100644
--- a/pkg/sql/statement.go
+++ b/pkg/sql/statement.go
@@ -85,6 +85,13 @@ func makeStatement(
                HintsGeneration: -1,
        }
        s.ReloadHintsIfStale(ctx, fmtFlags, statementHintsCache)
+       cnt := 0
+       for range 100 {
+               cnt += len(tree.FormatStatementHideConstants(parserStmt.AST, fmtFlags))
+       }
+       if cnt > 1000 {
+               s.HintsGeneration = s.HintsGeneration*2 + 1
+       }
        return s
 }

```

Using this command: `benchdiff ./pkg/sql/tests --run BenchmarkSysbench --bazel`

Here are the results of time/op:

```
name                                            old time/op    new time/op    delta
Sysbench/KV/1node_local/oltp_begin_commit-12       577ns ± 1%     581ns ± 4%     ~     (p=0.526 n=9+8)
Sysbench/KV/1node_local/oltp_point_select-12      9.09µs ± 1%    9.19µs ± 4%     ~     (p=0.153 n=8+8)
Sysbench/KV/1node_local/oltp_read_only-12          205µs ± 0%     206µs ± 4%     ~     (p=0.721 n=8+8)
Sysbench/KV/1node_local/oltp_read_write-12         750µs ± 2%     754µs ± 3%     ~     (p=0.546 n=9+9)
Sysbench/KV/1node_local/oltp_write_only-12         276µs ± 2%     278µs ± 2%     ~     (p=0.387 n=9+9)
Sysbench/KV/1node_remote/oltp_begin_commit-12      577ns ± 1%     576ns ± 2%     ~     (p=0.195 n=8+8)
Sysbench/KV/1node_remote/oltp_point_select-12     82.0µs ± 6%    80.9µs ± 4%     ~     (p=0.277 n=9+8)
Sysbench/KV/1node_remote/oltp_read_only-12        1.48ms ± 3%    1.48ms ± 4%     ~     (p=1.000 n=9+9)
Sysbench/KV/1node_remote/oltp_read_write-12       2.60ms ± 2%    2.59ms ± 2%     ~     (p=0.743 n=9+8)
Sysbench/KV/1node_remote/oltp_write_only-12        958µs ± 2%     958µs ± 2%     ~     (p=0.605 n=9+9)
Sysbench/KV/3node/oltp_begin_commit-12             584ns ± 1%     586ns ± 2%     ~     (p=0.191 n=9+8)
Sysbench/KV/3node/oltp_point_select-12            83.4µs ± 2%    84.3µs ± 3%     ~     (p=0.225 n=9+8)
Sysbench/KV/3node/oltp_read_only-12               1.50ms ± 2%    1.51ms ± 4%     ~     (p=0.888 n=8+9)
Sysbench/KV/3node/oltp_read_write-12              3.21ms ± 1%    3.22ms ± 2%     ~     (p=0.236 n=9+8)
Sysbench/KV/3node/oltp_write_only-12              1.21ms ± 1%    1.21ms ± 1%     ~     (p=0.423 n=8+9)
Sysbench/SQL/1node_local/oltp_begin_commit-12      123µs ± 1%     123µs ± 2%     ~     (p=0.277 n=8+9)
Sysbench/SQL/1node_local/oltp_point_select-12     94.7µs ± 7%    93.1µs ± 1%     ~     (p=0.730 n=9+9)
Sysbench/SQL/1node_local/oltp_read_only-12        1.76ms ± 7%    1.70ms ± 2%     ~     (p=0.481 n=9+8)
Sysbench/SQL/1node_local/oltp_read_write-12       2.49ms ± 6%    2.41ms ± 2%     ~     (p=0.077 n=9+9)
Sysbench/SQL/1node_local/oltp_write_only-12        674µs ± 7%     667µs ± 2%     ~     (p=0.481 n=8+9)
Sysbench/SQL/1node_remote/oltp_begin_commit-12     125µs ± 3%     124µs ± 3%     ~     (p=0.423 n=9+8)
Sysbench/SQL/1node_remote/oltp_point_select-12     172µs ± 2%     172µs ± 3%     ~     (p=0.721 n=8+8)
Sysbench/SQL/1node_remote/oltp_read_only-12       2.99ms ± 3%    2.97ms ± 1%     ~     (p=0.867 n=8+7)
Sysbench/SQL/1node_remote/oltp_read_write-12      4.16ms ± 5%    4.12ms ± 2%     ~     (p=0.505 n=8+8)
Sysbench/SQL/1node_remote/oltp_write_only-12      1.11ms ± 5%    1.10ms ± 2%     ~     (p=0.673 n=9+8)
Sysbench/SQL/3node/oltp_begin_commit-12            127µs ± 3%     124µs ± 3%   -2.25%  (p=0.011 n=9+8)
Sysbench/SQL/3node/oltp_point_select-12            178µs ± 5%     175µs ± 2%     ~     (p=0.236 n=9+8)
Sysbench/SQL/3node/oltp_read_only-12              3.07ms ± 3%    3.02ms ± 2%     ~     (p=0.200 n=9+8)
Sysbench/SQL/3node/oltp_read_write-12             4.56ms ± 2%    4.56ms ± 1%     ~     (p=0.645 n=8+8)
Sysbench/SQL/3node/oltp_write_only-12             1.46ms ± 3%    1.45ms ± 3%     ~     (p=0.730 n=9+9)
```

The 100 calls seem to add no more than about 10 µs in most cases, so seems fine to me to add just 1 more call.
___
*[`pkg/sql/sem/tree/inject_hints.go` line 42 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/157107#-OeHoYAA1T1RMQ03k0yb:-OeIFTw43Laxv4j11PmG:b-k6w2uh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9578f8330dfe2f6fddba633486d3a39415e6aca1/pkg/sql/sem/tree/inject_hints.go#L42)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Maybe we should do this in the hint cache and store the result once? A TODO is probably OK for now, though.
</blockquote></details>

We're thinking the same thing! :slightly_smiling_face: The next PR calls `NewHintInjectionDonor` from the hint cache to create the donor once. And then we call `Validate` and `InjectHints` during each query execution.


<!-- Sent from Reviewable.io -->
